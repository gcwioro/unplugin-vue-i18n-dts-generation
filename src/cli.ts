#!/usr/bin/env node

import path from "node:path";
import {parseArgs} from "node:util";
import {watch} from "chokidar";
import {generateI18nTypes} from "./api";
import type {GenerateTypesOptions} from "./api";

const helpText = `
unplugin-vue-i18n-dts-generation CLI

Generate TypeScript definitions from Vue i18n locale files

Usage:
  npx unplugin-vue-i18n-dts-generation generate [options]

Options:
  --root <path>              Root directory (default: current directory)
  --include <pattern>        Glob pattern(s) for locale files (can be specified multiple times)
  --exclude <pattern>        Glob pattern(s) to exclude (can be specified multiple times)
  --base-locale <locale>     Base locale for type generation (default: "de")
  --types-path <path>        Output path for .d.ts file (default: "./vite-env-override.d.ts")
  --virtual-file-path <path> Output path for virtual module .ts file (optional)
  --source-id <id>           Virtual module ID (default: "virtual:unplug-i18n-dts-generation")
  --banner <text>            Custom banner comment for generated files
  --merge <type>             Merge strategy: "deep" or "shallow" (default: "deep")
  --watch, -w                Watch mode - regenerate on file changes
  --debug                    Enable debug logging
  --verbose, -v              Enable verbose output
  --help, -h                 Show this help message
  --version                  Show version

Examples:
  # Basic usage with default settings
  npx unplugin-vue-i18n-dts-generation generate

  # Specify custom paths and base locale
  npx unplugin-vue-i18n-dts-generation generate --base-locale en --types-path src/types/i18n.d.ts

  # Multiple include patterns
  npx unplugin-vue-i18n-dts-generation generate --include "src/locales/**/*.json" --include "src/i18n/**/*.json"

  # Generate virtual file for debugging
  npx unplugin-vue-i18n-dts-generation generate --virtual-file-path src/i18n/virtual.gen.ts --verbose

  # Custom root directory
  npx unplugin-vue-i18n-dts-generation generate --root ./packages/frontend --base-locale en

  # Watch mode - regenerate on file changes
  npx unplugin-vue-i18n-dts-generation generate --watch --verbose
`;

async function main() {
  try {
    const args = process.argv.slice(2);

    // Handle help and version
    if (args.includes("--help") || args.includes("-h") || args.length === 0) {
      console.log(helpText);
      process.exit(0);
    }

    if (args.includes("--version")) {
      // Read version from package.json
      try {
        const {version} = await import("../package.json");
        console.log(`v${version}`);
      } catch {
        console.log("Version unavailable");
      }
      process.exit(0);
    }

    // Check if first arg is "generate" command
    let startIndex = 0;
    if (args[0] === "generate") {
      startIndex = 1;
    }

    // Parse arguments
    const {values} = parseArgs({
      args: args.slice(startIndex),
      options: {
        root: {type: "string"},
        include: {type: "string", multiple: true},
        exclude: {type: "string", multiple: true},
        "base-locale": {type: "string"},
        "types-path": {type: "string"},
        "virtual-file-path": {type: "string"},
        "source-id": {type: "string"},
        banner: {type: "string"},
        merge: {type: "string"},
        watch: {type: "boolean", short: "w"},
        debug: {type: "boolean"},
        verbose: {type: "boolean", short: "v"},
      },
      strict: true,
      allowPositionals: false,
    });

    // Build options
    const options: GenerateTypesOptions = {
      root: values.root,
      verbose: values.verbose ?? values.debug,
      debug: values.debug,
    };

    if (values.include) {
      options.include = values.include;
    }

    if (values.exclude) {
      options.exclude = values.exclude;
    }

    if (values["base-locale"]) {
      options.baseLocale = values["base-locale"];
    }

    if (values["types-path"]) {
      options.typesPath = values["types-path"];
    }

    if (values["virtual-file-path"]) {
      options.virtualFilePath = values["virtual-file-path"];
    }

    if (values["source-id"]) {
      options.sourceId = values["source-id"];
    }

    if (values.banner) {
      options.banner = values.banner;
    }

    if (values.merge) {
      if (values.merge !== "deep" && values.merge !== "shallow") {
        console.error(`‚ùå Error: --merge must be either "deep" or "shallow"`);
        process.exit(1);
      }
      options.merge = values.merge;
    }

    // Run generation
    let lastResult: Awaited<ReturnType<typeof generateI18nTypes>> | null = null;

    const runGeneration = async (): Promise<void> => {
      const startTime = Date.now();
      console.log("üöÄ Generating i18n types...\n");

      try {
        const result = await generateI18nTypes(options);
        lastResult = result;

        const duration = Date.now() - startTime;
        console.log(`\n‚úÖ Generation complete in ${duration}ms!`);
        console.log(`üìÅ Generated files (${result.filesWritten}):`);
        for (const file of result.generatedFiles) {
          console.log(`   - ${file}`);
        }
        console.log(`üåç Locales (${result.locales.length}): ${result.locales.join(", ")}`);
        console.log(`üìÑ Processed ${result.localeFilesCount} locale file(s)`);
      } catch (error) {
        console.error("\n‚ùå Generation failed:");
        if (error instanceof Error) {
          console.error(error.message);
          if (options.debug || options.verbose) {
            console.error("\nStack trace:");
            console.error(error.stack);
          }
        } else {
          console.error(String(error));
        }
        if (!values.watch) {
          process.exit(1);
        }
      }
    };

    // Initial generation
    await runGeneration();

    // Watch mode
    if (values.watch) {
      console.log("\nüëÅÔ∏è  Watch mode enabled - watching for changes...");
      console.log("Press Ctrl+C to exit\n");

      if (!lastResult) {
        console.error("‚ùå Cannot start watch mode: Initial generation failed");
        process.exit(1);
      }

      const rootDir = path.resolve(options.root || process.cwd());
      const includePatterns = options.include || ['src/**/locales/*.json'];

      if (options.verbose || options.debug) {
        console.log(`[watch] Root directory: ${rootDir}`);
        console.log(`[watch] Patterns: ${Array.isArray(includePatterns) ? includePatterns.join(', ') : includePatterns}`);
      }

      // Use the locale files from the initial generation
      const localeFilesFullPaths = lastResult?.localeFiles || [];

      if (options.verbose || options.debug) {
        console.log(`[watch] Found ${localeFilesFullPaths.length} locale file(s) to watch`);
      }

      // Watch the actual files that were found
      const watcher = watch(localeFilesFullPaths.length > 0 ? localeFilesFullPaths : includePatterns, {
        cwd: rootDir,
        ignored: options.exclude || ['**/node_modules/**', '**/.git/**'],
        persistent: true,
        ignoreInitial: true, // Skip initial add events since we've already generated
        awaitWriteFinish: {
          stabilityThreshold: 300,
          pollInterval: 100,
        },
      });

      // Debounce to avoid multiple rapid regenerations
      let debounceTimer: NodeJS.Timeout | null = null;
      const debounceDelay = 500; // 500ms debounce
      let isWatcherReady = false;

      const scheduleRegeneration = (eventType: string, filePath: string) => {
        // Skip initial 'add' events since we've already generated once
        if (!isWatcherReady && eventType === 'added') {
          return;
        }

        if (debounceTimer) {
          clearTimeout(debounceTimer);
        }

        debounceTimer = setTimeout(async () => {
          console.log(`\nüìù File ${eventType}: ${filePath}`);
          await runGeneration();
        }, debounceDelay);
      };

      watcher
        .on('ready', () => {
          isWatcherReady = true;
          const watched = watcher.getWatched();
          const watchedCount = Object.values(watched).reduce((sum, files) => sum + files.length, 0);

          if (options.verbose || options.debug) {
            console.log(`[watch] Watcher ready. Watching ${watchedCount} file(s):`);
            for (const [dir, files] of Object.entries(watched)) {
              if (files.length > 0) {
                console.log(`  ${dir}:`);
                for (const file of files) {
                  console.log(`    - ${file}`);
                }
              }
            }
          }

          if (watchedCount === 0) {
            console.warn(`\n‚ö†Ô∏è  Warning: No files are being watched!`);
            console.warn(`   Root: ${rootDir}`);
            console.warn(`   Patterns: ${Array.isArray(includePatterns) ? includePatterns.join(', ') : includePatterns}`);
            console.warn(`   This may indicate that the patterns don't match any files.`);
          }
        })
        .on('add', (path) => scheduleRegeneration('added', path))
        .on('change', (path) => scheduleRegeneration('changed', path))
        .on('unlink', (path) => scheduleRegeneration('removed', path))
        .on('error', (error) => {
          console.error('\n‚ùå Watcher error:', error);
        });

      // Keep process alive
      process.on('SIGINT', () => {
        console.log('\n\nüëã Stopping watch mode...');
        watcher.close();
        process.exit(0);
      });

      // Prevent exit
      return;
    }

    process.exit(0);
  } catch (error) {
    console.error("\n‚ùå Error generating types:");
    if (error instanceof Error) {
      console.error(error.message);
      if (process.argv.includes("--debug") || process.argv.includes("--verbose")) {
        console.error("\nStack trace:");
        console.error(error.stack);
      }
    } else {
      console.error(String(error));
    }
    process.exit(1);
  }
}

main();
