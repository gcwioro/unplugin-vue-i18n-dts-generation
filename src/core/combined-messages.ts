import {JSONObject} from "../types";
import {fnv1a32, getFinalKeys} from "../utils";

export class CombinedMessages<
  TLanguages extends string = string,
  TMessages extends JSONObject = JSONObject,
> {
  public readonly messagesJsonString: string;
  public readonly keys: string[];
  public readonly baseLocaleMessages: TMessages;
  public readonly languages: TLanguages[];
  public readonly contentId: string;
  public readonly baseLocale: TLanguages;

  public constructor(
    public readonly messages: Record<TLanguages, TMessages>,
    baseLocale: TLanguages,
  ) {
    this.messagesJsonString = JSON.stringify(messages);

    const languages = Object.keys(messages).filter((locale) => locale !== "js-reserved");
    this.languages = Array.from(new Set(languages)).sort((a, b) => (a < b ? -1 : a > b ? 1 : 0)) as TLanguages[];

    const resolvedBaseLocale = (this.languages.includes(baseLocale)
      ? baseLocale
      : this.languages[0]) as TLanguages | undefined;

    this.baseLocale = resolvedBaseLocale ?? baseLocale;

    const fallbackLocale = this.baseLocale in messages ? this.baseLocale : (this.languages[0] as TLanguages | undefined);
    this.baseLocaleMessages = (fallbackLocale ? messages[fallbackLocale] : undefined) ?? ({} as TMessages);

    const keysLocale = fallbackLocale ?? this.baseLocale;
    this.keys = getFinalKeys(messages, keysLocale as keyof typeof messages);

    this.contentId = fnv1a32(this.messagesJsonString);
  }

  public languagesTuple(): string {
    return `['${this.languages.join(`', '`)}']`;
  }
}
