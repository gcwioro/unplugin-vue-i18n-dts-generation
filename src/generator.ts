import type {DtsContentParams, JSONObject, JSONValue} from './types'
import {getJsonLeafPaths} from './utils/json'
import {transformKeys} from './utils/keys'
import {fnv1a32} from './utils/hash'

/**
 * Create .d.ts content for type definitions only
 */
export function toTypesContent(params: Omit<DtsContentParams, 'typeFilePath'>): string {
  const {
    baseLocale,
    messages,
    supportedLanguages,
    banner,
    sourceId = '@unplug-i18n-types-locales',
  } = params

  // Derive leaf paths
  const messagesForBaseLocale = (messages?.[baseLocale] ?? Object.values(messages)[0]) as Record<string, JSONValue>;
  const allKeysRaw = getJsonLeafPaths(messagesForBaseLocale)
    // Remove trailing ".body.cases" (Vue I18n 9.x specific)
    .map((p) => p.replace(/\.body\.cases$/g, ''))

  const finalKeys = transformKeys(allKeysRaw)

  // Messages are already canonicalized from plugin.ts
  const canonicalBase = messages as JSONObject
  const baseLocaleMessages = canonicalBase[baseLocale] ?? Object.values(canonicalBase)[0]

  // Deterministic banner (no timestamps). Include a small content hash for traceability.
  const messagesJson = JSON.stringify(canonicalBase)
  const languagesTuple = `['${supportedLanguages.filter(l=>l!= 'js-reserved').join(`', '`)}']`
  const contentId = fnv1a32(languagesTuple + '|' + finalKeys.join('|') + '|' + messagesJson)

  const autogenerated =
    (banner ?? //// @ts-nocheck
      `/* eslint-disable */
       /* prettier-ignore */
      // biome-ignore lint: disable
      // noinspection JSUnusedGlobalSymbols'
      // AUTO-GENERATED FILE. DO NOT EDIT.
      // Content-Hash: ${contentId}
`.split('\n').map(l => l.trim()).filter(l => l).join('\n') + '\n\n')

  const NL = '\n'

  // Type definitions only - build with array for better performance
  const bodyLines = [
    '',
    `declare module '${sourceId}' {`,
    '',
    '',
    '  const messages: Record<string, any>',
    '  export default messages',
    '}',
    '',
    "import type { I18nOptions } from 'vue-i18n'",
    '',
    `export type AllTranslationKeysGen = ${finalKeys.length ? `'${finalKeys.join(`' | '`)}'` : 'never'}`,
    `export type SupportedLanguagesGen = readonly [${supportedLanguages.filter(l => l != 'js-reserved').map(l => `'${l}'`).join(', ')}]`,
    'export type SupportedLanguageUnionGen = SupportedLanguagesGen[number]',
    '',
    '// Message structure types',
    `export type MessageSchemaGen = ${JSON.stringify(baseLocaleMessages, null, 2)}`,
    'export type AllLocaleGen = Readonly<Record<SupportedLanguageUnionGen, MessageSchemaGen>>',
    'export type AllTranslationsGen = AllLocaleGen',
    '',
    '// Type-safe translate function parameters',
    'export type TranslateParams = (string | number | undefined | null) | Record<string, unknown>',
    '',
    '// I18n config options (excludes messages as they\'re provided by the plugin)',
    'export type I18nConfigOptions = Omit<I18nOptions<MessageSchemaGen, {}, {}, SupportedLanguageUnionGen, false>, \'messages\'>',
    ''
  ];

  return (autogenerated + bodyLines.join(NL)).replace(/\r\n/g, NL).replace(/\r/g, NL).replace(/\n*$/, '\n')
}

/**
 * Create .ts content for const values
 */
export function toConstsContent(params: DtsContentParams): string {
  const {
    baseLocale,
    messages,
    typeFilePath,
    supportedLanguages,
    banner,
    exportMessages = true,
    sourceId = '@unplug-i18n-types-locales'
  } = params

  // Messages are already canonicalized from plugin.ts
  const canonicalBase = messages as JSONObject

  // Deterministic banner
  const messagesJson = JSON.stringify(canonicalBase)
  const languagesTuple = `['${supportedLanguages.filter(l=>l!= 'js-reserved').join(`', '`)}']`
  const contentId = fnv1a32(languagesTuple + '|' + messagesJson)

  const autogenerated =
    (banner ??
      `// AUTO-GENERATED FILE. DO NOT EDIT.
// Generated by unplugin-vue-i18n-dts-generation (consts)
// Content-Hash: ${contentId}
`)

  const NL = '\n'

  // Runtime const values - build with array for better performance
  const allKeys = exportMessages ? transformKeys(getJsonLeafPaths((messages?.[baseLocale] ?? Object.values(messages)[0]) as Record<string, JSONValue>).map((p) => p.replace(/\.body\.cases$/g, ''))) : [];

  const bodyLines = [
    '',
    "import type {I18n, I18nOptions} from 'vue-i18n'",
    "import { type Plugin} from 'vue'",
    'import {',
    '  type ComposerOptions,',
    '  createI18n,',
    '  type Locale,',
    '  type NamedValue,',
    '  type TranslateOptions,',
    '  useI18n,',
    '  type UseI18nOptions',
    "} from 'vue-i18n'",
    '// @ts-expect-error - Virtual module provided by vite-plugin-locale-json',
    `import _messagesI18n from '${sourceId}'`,
    'import type {',
    '  AllTranslationKeysGen,',
    '  AllTranslationsGen,',
    '  SupportedLanguagesGen,',
    `} from '${typeFilePath.replace(/\.d\.ts$/, '')}'`,
    '',
    `export const supportedLanguages = ${languagesTuple} as const satisfies SupportedLanguagesGen`,
    exportMessages ? `export const messages = ${messagesJson} as const` : '',
    '',
    '// Type-safe i18n messages',
    'export const messagesI18n = _messagesI18n as unknown as AllTranslationsGen',
    '',
    '',
  ];

  if (exportMessages) {
    bodyLines.push('// Helper to get all translation keys');
    bodyLines.push('export const allTranslationKeys: AllTranslationKeysGen[] = [');
    bodyLines.push(...allKeys.map(k => `  '${k}',`));
    bodyLines.push(']');
    bodyLines.push('');
  }

  bodyLines.push(
    '',
    '',
    '/**',
    ' * Creates a type-safe i18n instance with pre-configured messages',
    ' * @param options - Optional I18n configuration (messages will be automatically provided)',
    ' * @returns Configured i18n instance',
    ' */',
    'export function createI18nInstance<T extends Partial<ComposerOptions> >(options?: T): I18n<AllTranslationsGen, T["datetimeFormats"] extends Record<string, unknown> ? T["datetimeFormats"] : {}, T["numberFormats"] extends Record<string, unknown> ? T["numberFormats"] : {}, T["locale"] extends string ? T["locale"] : Locale, false> {',
    '  const i18Options = {',
    `    fallbackLocale: '${baseLocale}',`,
    '    missingWarn: false,',
    '    fallbackWarn: false,',
    '    ...options,',
    '    // Messages are provided and cannot be overridden',
    '    messages: messagesI18n,',
    '    legacy: false',
    '  } as const;',
    '  const i18n = createI18n<false, typeof i18Options, AllTranslationsGen>(i18Options);',
    '  return i18n;',
    '}',
    'export function createI18nInstancePlugin<T extends Partial<ComposerOptions>&I18nOptions >(options?: T): Plugin<unknown[]>&( I18n<AllTranslationsGen, T["datetimeFormats"] extends Record<string,unknown> ? T["datetimeFormats"] : {}, T["numberFormats"] extends Record<string, unknown> ? T["numberFormats"] : {}, T["locale"] extends string ? T["locale"] : Locale, false> ){',
    '  const i18n = createI18nInstance(options);',
    '  return i18n as any;',
    '}',
    'export declare interface I18nCustom {',
    '  (key: AllTranslationKeysGen, plural: number, options?: TranslateOptions): string',
    '',
    '  (key: AllTranslationKeysGen, options?: TranslateOptions): string',
    '',
    '  (key: AllTranslationKeysGen, defaultMsg?: string): string',
    '',
    '  (key: AllTranslationKeysGen, defaultMsg: string, options?: TranslateOptions): string',
    '',
    '  (key: AllTranslationKeysGen, named: NamedValue, defaultMsg?: string): string',
    '',
    '  (key: AllTranslationKeysGen, named: NamedValue, plural?: number): string',
    '',
    '  (key: AllTranslationKeysGen, named: NamedValue, options?: TranslateOptions): string',
    '  (key: AllTranslationKeysGen, plural: number, named: NamedValue): string',
    '  (key: AllTranslationKeysGen, plural: number, defaultMsg: string): string',
    '}',
    '',
    '',
    "export function useI18nTypeSafe(options?: Omit<UseI18nOptions, 'messages'>) {",
    '  // const instance = i18n || createI18nInstance()',
    '  const {t: originalT, d, n, locale, ...rest} =',
    '    useI18n(Object.assign(options ?? {',
    `      fallbackLocale: '${baseLocale}'`,
    '    }, {messages: messagesI18n}))',
    '',
    '  const t: I18nCustom = originalT as I18nCustom satisfies I18nCustom',
    '  return {',
    '    ...rest,',
    '    t,',
    '    d,',
    '    n,',
    '    locale,',
    '    availableLocales: supportedLanguages as unknown as string[],',
    '    // i18n: instance',
    '  }',
    '}',
    ''
  );

  return (autogenerated + bodyLines.join(NL)).replace(/\r\n/g, NL).replace(/\r/g, NL).replace(/\n*$/, '\n')
}

/**
 * Emit **pure JS** (no TS annotations, no `as const`).
 * Rollup will replace `import.meta.ROLLUP_FILE_URL_<ref>` in build.
 */
export function createVirtualModuleCode(opts: {
  jsonText: string;
  exportData: boolean;
  buildAssetRefId?: string;
  devUrlPath?: string;
}) {
  const {jsonText, exportData, buildAssetRefId, devUrlPath} = opts;

  const urlExpr = buildAssetRefId
    ? `import.meta.ROLLUP_FILE_URL_${buildAssetRefId}`
    : JSON.stringify(devUrlPath || "/_virtual_locales.json");

  // Pure JS. If you want types, add an ambient .d.ts for "virtual:locales".
  if (exportData) {
    // Inline the data directly in the module
    return `
const url = ${urlExpr};

export async function load() {
  const res = await fetch(url, { credentials: 'same-origin' });
  if (!res.ok) throw new Error('Failed to load locales: ' + res.status);
  return await res.json();
}

export default ${jsonText};
`;
  } else {
    // Don't inline the data, only provide the load function
    return `
const url = ${urlExpr};

export async function load() {
  const res = await fetch(url, { credentials: 'same-origin' });
  if (!res.ok) throw new Error('Failed to load locales: ' + res.status);
  return await res.json();
}

export default undefined;
`;
  }
}
