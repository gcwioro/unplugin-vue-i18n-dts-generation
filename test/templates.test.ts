import {describe, expect, it} from 'vitest'
import {CombinedMessages} from '../src/core/combined-messages'
import {generateModuleArtifacts} from '../src/generator'

const sampleMessages = {
  en: {
    greeting: 'Hello',
    nested: {
      welcome: 'Welcome',
    },
  },
  de: {
    greeting: 'Hallo',
    nested: {
      welcome: 'Willkommen',
    },
  },
}

describe('template rendering', () => {
  it('renders messages and runtime modules', () => {
    const combined = new CombinedMessages(sampleMessages, 'en')
    const artifacts = generateModuleArtifacts({combinedMessages: combined})

    expect(artifacts.messages.ts).toMatchSnapshot('messages-ts')
    expect(artifacts.runtime.ts).toMatchSnapshot('runtime-ts')
    expect(artifacts.typesContent).toMatchSnapshot('types-dts')
  })
})
