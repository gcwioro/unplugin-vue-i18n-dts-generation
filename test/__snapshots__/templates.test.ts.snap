// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template rendering > renders messages and runtime modules > messages-ts 1`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// AUTO-GENERATED FILE. DO NOT EDIT.
// Content-Hash: b1586b08
export const baseLocale = "en" as const

export const messages = {
  "en": {
    "greeting": "Hello",
    "nested": {
      "welcome": "Welcome"
    }
  },
  "de": {
    "greeting": "Hallo",
    "nested": {
      "welcome": "Willkommen"
    }
  }
} as const

export const supportedLanguages = ['de', 'en'] as const

export type AllSupportedLanguages = readonly ['de', 'en']
export type SupportedLanguage = AllSupportedLanguages[number]

export type AllTranslationKeys = 'greeting' | 'nested.welcome'

export type MessageSchemaGen = (typeof messages)[typeof baseLocale]
export type AllTranslations = typeof messages
export type MessagesType = AllTranslations

export default messages
"
`;

exports[`template rendering > renders messages and runtime modules > runtime-ts 1`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// AUTO-GENERATED FILE. DO NOT EDIT.
// Content-Hash: b1586b08
import {createI18n, useI18n} from 'vue-i18n'
import type {Composer, DefineDateTimeFormat, DefineNumberFormat, FallbackLocale, I18n, I18nOptions, Locale, NamedValue, TranslateOptions, UseI18nOptions} from 'vue-i18n'
import messages, {baseLocale, supportedLanguages} from './messages.ts'
import type {AllTranslationKeys, MessageSchemaGen, MessagesType, SupportedLanguage} from './messages.ts'

const FALLBACK_LOCALES = supportedLanguages.reduce<Record<SupportedLanguage, SupportedLanguage[]>>((acc: Record<SupportedLanguage, SupportedLanguage[]>, locale: SupportedLanguage) => {
  const chain = new Set<SupportedLanguage>()
  chain.add(locale)
  if (locale !== "en") {
    chain.add("en" as SupportedLanguage)
  }
  acc[locale] = Array.from(chain)
  return acc
}, {} as Record<SupportedLanguage, SupportedLanguage[]>)

type WritableMessages = {
  -readonly [Locale in keyof typeof messages]: {
    -readonly [Key in keyof typeof messages[Locale]]: typeof messages[Locale][Key];
  };
};

const mutableMessages = messages as WritableMessages;

function withNumericSecondArg<Fn extends (...args: any[]) => any>(fn: Fn): Fn {
  return ((...args: Parameters<Fn>) => {
    const nextArgs = [...args] as unknown[]
    if (nextArgs.length >= 2) {
      const second = nextArgs[1]
      const numeric = typeof second === 'number'
        ? second
        : typeof second === 'string'
          ? Number.parseFloat(second)
          : Number.NaN
      if (Number.isFinite(numeric)) {
        const third = nextArgs[2]
        const named = third && typeof third === 'object' && !Array.isArray(third)
          ? {...third as Record<string, unknown>}
          : {}
        named.count = numeric
        named.n = numeric
        nextArgs[2] = named
      }
    }
    return fn(...nextArgs as Parameters<Fn>) as ReturnType<Fn>
  }) as Fn
}

export type TranslateParams = (string | number | undefined | null) | Record<string, unknown>;

export interface I18nCustom {
  (key: AllTranslationKeys, plural: number, options?: TranslateOptions): string;
  (key: AllTranslationKeys, options?: TranslateOptions): string;
  (key: AllTranslationKeys, defaultMsg?: string): string;
  (key: AllTranslationKeys, defaultMsg: string, options?: TranslateOptions): string;
  (key: AllTranslationKeys, named: NamedValue, defaultMsg?: string): string;
  (key: AllTranslationKeys, named: NamedValue, plural?: number): string;
  (key: AllTranslationKeys, named: NamedValue, options?: TranslateOptions): string;
  (key: AllTranslationKeys, plural: number, named: NamedValue): string;
  (key: AllTranslationKeys, plural: number, defaultMsg: string): string;
}

type DateTimeSchema = Record<string, DefineDateTimeFormat>;
type NumberSchema = Record<string, DefineNumberFormat>;
type SchemaDefinition = {
  message: MessageSchemaGen;
  datetime: DateTimeSchema;
  number: NumberSchema;
};
type DateTimeFormatsMap = Record<SupportedLanguage, DateTimeSchema>;
type NumberFormatsMap = Record<SupportedLanguage, NumberSchema>;

type VueI18nInstance = I18n<MessagesType, DateTimeFormatsMap, NumberFormatsMap, SupportedLanguage, false>;
type ComposerInstance = Composer<MessagesType, DateTimeFormatsMap, NumberFormatsMap, SupportedLanguage>;

export type I18nConfigOptions = I18nOptions<SchemaDefinition, SupportedLanguage>;
type UseI18nSchema = UseI18nOptions<SchemaDefinition, SupportedLanguage>;

export type UseI18nTypesafeReturn = Omit<ComposerInstance, 't'> & { t: I18nCustom };

export function createI18nInstance<T extends Partial<I18nConfigOptions>>(options?: T) {
  const locale = options?.locale ?? (typeof navigator !== 'undefined' && navigator.language ? navigator.language : baseLocale);
  const typedOptions: I18nConfigOptions = {
    legacy: false,
    fallbackLocale: FALLBACK_LOCALES as FallbackLocale,
    locale: locale as SupportedLanguage,
    messages,
    ...(options ?? {}),
  };
  const i18n = createI18n(typedOptions as I18nConfigOptions & { legacy: false; messages: MessagesType });
  const instance = i18n as unknown as VueI18nInstance;
  (globalThis as Record<string, unknown>).i18nApp = instance;
  return instance;
}

export function createI18nInstancePlugin<T extends Partial<I18nConfigOptions>>(options?: T): VueI18nInstance {
  return createI18nInstance(options);
}

export function useI18nTypeSafe(options?: Omit<UseI18nSchema, 'messages'>): UseI18nTypesafeReturn {
  const composer = useI18n({
    ...(options ?? {}),
    messages,
  } as UseI18nSchema) as unknown as ComposerInstance;
  const {t, ...rest} = composer;
  return {
    ...rest,
    t: withNumericSecondArg(t),
  };
}

export async function load() {
  return messages
}

export {messages, supportedLanguages};
export {baseLocale};

type ImportMetaHot = {
  hot?: {
    on: (event: string, cb: (payload: {messages: Partial<typeof messages>}) => void) => void;
  };
};

function applyMessagesUpdate(next: Partial<typeof messages>) {
  for (const [locale, value] of Object.entries(next)) {
    if (!value) continue
    ;(mutableMessages as Record<string, unknown>)[locale] = value
  }
}

const hot = (import.meta as ImportMetaHot).hot
if (hot) {
  hot.on('i18n-update', (data: {messages: Partial<typeof messages>}) => {
    applyMessagesUpdate(data.messages)
    const instance = (globalThis as Record<string, unknown>).i18nApp as
      | VueI18nInstance
      | undefined
    if (instance) {
      for (const [locale, value] of Object.entries(data.messages)) {
        if (value) {
          instance.global.setLocaleMessage(locale, value as MessageSchemaGen)
        }
      }
      const current = instance.global.locale.value
      instance.global.locale.value = current
    }
  })
}

export default messages;
"
`;

exports[`template rendering > renders messages and runtime modules > types-dts 1`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// AUTO-GENERATED FILE. DO NOT EDIT.
// Content-Hash: b1586b08

import type { Composer, DefineDateTimeFormat, DefineNumberFormat, I18n, I18nOptions, NamedValue, TranslateOptions, UseI18nOptions } from 'vue-i18n';
import messages, { baseLocale, supportedLanguages } from './messages.js';
import type { AllTranslationKeys, MessageSchemaGen, MessagesType, SupportedLanguage } from './messages.js';

declare module 'virtual:vue-i18n-types' {
  export type TranslateParams = (string | number | undefined | null) | Record<string, unknown>;
  export interface I18nCustom {
      (key: AllTranslationKeys, plural: number, options?: TranslateOptions): string;
      (key: AllTranslationKeys, options?: TranslateOptions): string;
      (key: AllTranslationKeys, defaultMsg?: string): string;
      (key: AllTranslationKeys, defaultMsg: string, options?: TranslateOptions): string;
      (key: AllTranslationKeys, named: NamedValue, defaultMsg?: string): string;
      (key: AllTranslationKeys, named: NamedValue, plural?: number): string;
      (key: AllTranslationKeys, named: NamedValue, options?: TranslateOptions): string;
      (key: AllTranslationKeys, plural: number, named: NamedValue): string;
      (key: AllTranslationKeys, plural: number, defaultMsg: string): string;
  }
  type DateTimeSchema = Record<string, DefineDateTimeFormat>;
  type NumberSchema = Record<string, DefineNumberFormat>;
  type SchemaDefinition = {
      message: MessageSchemaGen;
      datetime: DateTimeSchema;
      number: NumberSchema;
  };
  type DateTimeFormatsMap = Record<SupportedLanguage, DateTimeSchema>;
  type NumberFormatsMap = Record<SupportedLanguage, NumberSchema>;
  type VueI18nInstance = I18n<MessagesType, DateTimeFormatsMap, NumberFormatsMap, SupportedLanguage, false>;
  type ComposerInstance = Composer<MessagesType, DateTimeFormatsMap, NumberFormatsMap, SupportedLanguage>;
  export type I18nConfigOptions = I18nOptions<SchemaDefinition, SupportedLanguage>;
  type UseI18nSchema = UseI18nOptions<SchemaDefinition, SupportedLanguage>;
  export type UseI18nTypesafeReturn = Omit<ComposerInstance, 't'> & {
      t: I18nCustom;
  };
  export function createI18nInstance<T extends Partial<I18nConfigOptions>>(options?: T): VueI18nInstance;
  export function createI18nInstancePlugin<T extends Partial<I18nConfigOptions>>(options?: T): VueI18nInstance;
  export function useI18nTypeSafe(options?: Omit<UseI18nSchema, 'messages'>): UseI18nTypesafeReturn;
  export function load(): Promise<{
      readonly en: {
          readonly greeting: "Hello";
          readonly nested: {
              readonly welcome: "Welcome";
          };
      };
      readonly de: {
          readonly greeting: "Hallo";
          readonly nested: {
              readonly welcome: "Willkommen";
          };
      };
  }>;
  export { messages, supportedLanguages };
  export { baseLocale };
  export default messages;

}

declare module 'virtual:vue-i18n-types/messages' {
  export const baseLocale: "en";
  export const messages: {
      readonly en: {
          readonly greeting: "Hello";
          readonly nested: {
              readonly welcome: "Welcome";
          };
      };
      readonly de: {
          readonly greeting: "Hallo";
          readonly nested: {
              readonly welcome: "Willkommen";
          };
      };
  };
  export const supportedLanguages: readonly ["de", "en"];
  export type AllSupportedLanguages = readonly ['de', 'en'];
  export type SupportedLanguage = AllSupportedLanguages[number];
  export type AllTranslationKeys = 'greeting' | 'nested.welcome';
  export type MessageSchemaGen = (typeof messages)[typeof baseLocale];
  export type AllTranslations = typeof messages;
  export type MessagesType = AllTranslations;
  export default messages;

}
"
`;
